### PR TITLE
Implement Btrieve ACS

### PR DESCRIPTION
**Pull request by paladine**
- Implements Btrieve ACS functionality + adds relevant tests
- Adds a version number to the Btrieve database for schema compatibility
- Speeds up unit tests by using a shared in memory account db for unit tests (not just integration)
- Handles database `Stream` reads to make sure we read the entire `Stream`

### CODE REVIEW

**Comment by paladine** [comment]
- Also added a `Seek` cursor direction for `KeyEqual` queries, instead of using that hack `>=` query + manual key validation. Manual key validation would have been more difficult given ACS. Makes sense to just let sqlite do it

File: MBBSEmu/Btrieve/BtrieveKey.cs [hunk 1]
```diff
@@ -61,6 +61,16 @@ public bool IsComposite
         /// </summary>
         public bool IsNullable { get => PrimarySegment.IsNullable; }
 
+        /// <summary>
+        ///     Whether this key requires ACS.
+        /// </summary>
+        public bool RequiresACS { get => Segments.Any(segment => segment.RequiresACS); }
```
Comment (enusbaum) [inline MBBSEmu/Btrieve/BtrieveKey.cs:67]: same as below

File: MBBSEmu/Btrieve/BtrieveUtil.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
+using System;
+using System.IO;
+
+namespace MBBSEmu.Btrieve
+{
+    public class BtrieveUtil
+    {
+        public static byte[] ReadEntireStream(Stream s)
+        {
+            int totalRead = 0;
```
Comment (enusbaum) [inline MBBSEmu/Btrieve/BtrieveUtil.cs:10]: var

File: MBBSEmu/Btrieve/BtrieveUtil.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            byte[] buffer = new byte[s.Length];
```
Comment (enusbaum) [inline MBBSEmu/Btrieve/BtrieveUtil.cs:11]: var

**Review by enusbaum — CHANGES_REQUESTED** [review]
formatting changes

File: MBBSEmu.Tests/Btrieve/BtrieveKey_Tests.cs [hunk 4]
```diff
@@ -223,5 +219,137 @@ public void NullValueString(EnumKeyDataType dataType)
...
+
+            return acs;
+        }
+
+        [Fact]
+        public void ACSReplacement_SingleKey()
+        {
+            var acs = UpperACS();
+
+            var key = new BtrieveKey() {
```
[the remark this answers is not in the document]
Comment (paladine) [reply MBBSEmu.Tests/Btrieve/BtrieveKey_Tests.cs:240]: TIL

**Pull request merged**